### PR TITLE
Adding publish to the oo_azure module.

### PR DESCRIPTION
**Pull request by kwoodson**
Adding the publish API call so we can publish our azure images.

@jim-minter 
We need to revisit our create_offer.yml so that it will cancel any ongoing operations so that the publish occurs.  I'll investigate this.

### CODE REVIEW

**Comment by jim-minter** [comment]
looks great so far

**Comment by vrutkovs** [comment]
bot, retest this please

**Comment by kwoodson** [comment]
@jim-minter, I have updated our `create_offer.yml` to support cancel and publish.  The issue I ran into during testing was that the cancel operation call takes a bit of time( 1-3 minutes) to cancel the previous publish.  Once it was cancelled then I was able to successfully publish again.

Should we poll their API for operation status to verify it completed?

File: playbooks/azure/openshift-cluster/create_offer.yml [hunk 1]
```diff
@@ -76,8 +76,20 @@
         msg: "{{ yeditout }}"
         verbosity: 1
 
-  - name: create an offer in cloudpartner portal
+    # this cancel operation returns a 202 whether it cancelled or not.
+    - name: cancel publish operation
```
Comment (jim-minter) [inline playbooks/azure/openshift-cluster/create_offer.yml]: indent wrong?

Comment (jim-minter) [reply playbooks/azure/openshift-cluster/create_offer.yml]: also I think I'd rename the playbook filename, it's doing much more than create at this point

File: playbooks/azure/openshift-cluster/create_offer.yml [hunk 2]
```diff
@@ -76,8 +76,20 @@
...
     oo_azure_rm_publish_image:
       offer: "{{ azure_image_publish[image_type].offer }}"
       offer_data: "{{ (lookup('template', 'offer.yml.j2') | from_yaml) if 'skipped' in yeditout and yeditout.skipped or not yeditout.changed else yeditout.results[0].result[0].edit }}"
       force: "{{ openshift_azure_offer_force |default(False)}}"
+
+  - name: publish this offer
+    oo_azure_rm_publish_image:
+      state: publish
+      offer: "{{ azure_image_publish[image_type].offer }}"
+      emails: "{{ openshift_azure_image_publish_emails }}"
```
Comment (jim-minter) [inline playbooks/azure/openshift-cluster/create_offer.yml]: creep on the number of input variables required - please document these

**Comment by jim-minter** [comment]
> Should we poll their API for operation status to verify it completed?

@kwoodson the API contract says you must do that.  Looks like you follow the Operation-Location header and check for status: failed | completed in the response?

File: playbooks/azure/openshift-cluster/create_offer.yml (hunk 1 above)
Comment (kwoodson) [reply playbooks/azure/openshift-cluster/create_offer.yml]: renamed.

**Comment by kwoodson** [comment]
@jim-minter, I have updated the publishing to poll the API until the operation is `canceled`.  I have tested this with a  limited number of tests.

File: roles/lib_utils/library/oo_azure_rm_publish_image.py [hunk 3]
```diff
@@ -160,6 +174,26 @@ def prepare_action(self, url, action='GET', data=None, add_headers=None):
 
         return AzurePublisher.request(action.upper(), self.server + url, data, headers)
 
+    def cancel_and_wait_for_operation(self, params):
+        '''cancel the current publish operation and wait for operation to complete'''
+
+        now = time.time()
+        # we need to wait here for 'submissionState' to move to 'canceled'
+        while True:
```
Comment (jim-minter) [inline roles/lib_utils/library/oo_azure_rm_publish_image.py]: shouldn't the self.cancel_operation should happen once before the loop?
is the timeout necessary?

File: roles/lib_utils/library/oo_azure_rm_publish_image.py [hunk 4]
```diff
@@ -194,19 +228,28 @@ def run_ansible(params):
...
         elif params['state'] == 'go_live':
             results = apc.go_live(offer=params['offer'])
         else:
-            raise AzurePublisherException('Unsupported query type: {}'.format(params['query']))
+            raise AzurePublisherException('Unsupported query type: {}'.format(params['state']))
 
         changed = False
 
-        if results.status_code in [200, 201]:
+        if results.status_code in [200, 201, 202]:
```
Comment (jim-minter) [inline roles/lib_utils/library/oo_azure_rm_publish_image.py]: change still necessary?

File: roles/lib_utils/library/oo_azure_rm_publish_image.py [hunk 5]
```diff
@@ -194,19 +228,28 @@ def run_ansible(params):
...
             changed = True
 
-        return {'data': results.json(), 'changed': changed, 'status_code': results.status_code}
+        rval = None
```
Comment (jim-minter) [inline roles/lib_utils/library/oo_azure_rm_publish_image.py]: under what circumstances is this change necessary, out of interest?

File: playbooks/azure/openshift-cluster/create_offer.yml (hunk 2 above)
Comment (jim-minter) [reply playbooks/azure/openshift-cluster/create_offer.yml]: thanks

**Comment by jim-minter** [comment]
@kwoodson I'm also uncomfortable about
```
         if (results.status_code == 200 or results.status_code == 404) and params['force']:
             return self.create_or_modify_offer(offer=params['offer'], data=params['offer_data'], modify=True)
 
         return self.create_or_modify_offer(offer=params['offer'], data=params['offer_data'])
```

the if statement doesn't look right to me - and I'm wondering why the force param even exists?

File: playbooks/azure/openshift-cluster/create_offer.yml (hunk 1 above)
Comment (kwoodson) [reply playbooks/azure/openshift-cluster/create_offer.yml]: This is a continuation of the block above.  I believe this is correct.

File: playbooks/azure/openshift-cluster/create_and_publish_offer.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+# Publish image
+
+The steps to build the image are as follows:
+
+## Step 1:
+
+Build the Openshift image using the build_node_image.yml playbook.  Once this playbook completes it should
+produce a stroage blob that points to the image.  This blob exists inside of the resourcegroup named images,
+storage accounts named openshiftimages, and the container named, blobs.
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply playbooks/azure/openshift-cluster/create_and_publish_offer.md]: thanks, updated.

File: roles/lib_utils/library/oo_azure_rm_publish_image.py (hunk 3 above)
Comment (kwoodson) [reply roles/lib_utils/library/oo_azure_rm_publish_image.py]: I could call cancel first or I could get the current `submissionState` and verify if it is `canceled` already.  I ran into a few race conditions where I could call publish/cancel and the get_operations would return that it was in state `running`.  If someone has previously canceled then this code doesn't require another cancel to be issued.

Thoughts?

File: roles/lib_utils/library/oo_azure_rm_publish_image.py (hunk 4 above)
Comment (kwoodson) [reply roles/lib_utils/library/oo_azure_rm_publish_image.py]: Good question.  If we return the cancel_operation then it was a `202`.  We are now returning the get_operation call which is `200`.  This doesn't hurt anything. 

What would you prefer?

File: roles/lib_utils/library/oo_azure_rm_publish_image.py (hunk 5 above)
Comment (kwoodson) [reply roles/lib_utils/library/oo_azure_rm_publish_image.py]: When calling `cancel_operation` there is no data returned in the results.  When calling `results.json()` an exception is thrown since there is nothing returned.  This exception handling allows us to handle emtpy json responses as long as the `results.ok` is `True`.

File: playbooks/azure/openshift-cluster/create_offer.yml (hunk 1 above)
Comment (kwoodson) [reply playbooks/azure/openshift-cluster/create_offer.yml]: I believe it is correct as it is part of the previous block. The block states that if there is an offer already there then run these steps which includes a cancel operation. If the offer does not exist then there is no need to run the cancel operation.

File: playbooks/azure/openshift-cluster/create_and_publish_offer.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+```
+$ ansible-playbook  create_and_publish_offer.yml -e @publishingvars.yml
+```
+
+Example publishingvars.yml
+```
+openshift_azure_container: images
+openshift_azure_storage_account: openshiftimages
+image_name: rhel7-3.9-201805211419
+openshift_azure_offer_force:  True
```
[the remark this answers is not in the document]
Comment (kwoodson) [reply playbooks/azure/openshift-cluster/create_and_publish_offer.md]: Good question here.  The docs for create or modify an offer explain that you need to pass the `If-Match` header if you want to update.  I added this flag in order for someone to allow the update.  If you attempt to update and its there then it throws an exception.

I can just assume `True` if you'd prefer.

**Comment by kwoodson** [comment]
@jim-minter, 
Changes:
- since the publish operation returns a 202 I need to keep it
- removed the timeout
- removed the force from the documentation and playbook
- call cancel outside the while loop and poll inside
- removed simplejson and exception handling for `.json()`
- added a lambda to return `''` for publish

**Comment by kwoodson** [comment]
/test gcp-upgrade

unrelated.

**Comment by jim-minter** [comment]
/lgtm

@sdodson or @michaelgugino please would you approve?
